File: 프로그래머스/1/178871. 달리기 경주/달리기 경주.js
function solution(players, callings) {
    const rankMap = new Map(Array.from(players, (x, i) => [x, i]));
    
    for (let name_current of callings) {
        const rank_current = rankMap.get(name_current);
        const rank_target = rank_current - 1;
        const name_target = players[rank_target];
        
        rankMap.set(name_current, rank_target);
        rankMap.set(name_target, rank_current);
        [players[rank_current], players[rank_target]] = [players[rank_target], players[rank_current]];
    }
    
    return players;
}

File: 프로그래머스/1/178871. 달리기 경주/달리기 경주.py
def solution(players, callings):
    rank = { x: i for (i, x) in enumerate(players) }
    
    for call in callings:
        idx = rank[call]
        rank[call] -= 1
        rank[players[idx-1]] += 1
        players[idx], players[idx-1] = players[idx-1], players[idx]
        
    return players